### DISCUSSION TITLE
adding more constraints in Constrained, Parallel, Multi-Objective BO in BoTorch with qEHVI and qParEGO

### DISCUSSION THREAD

**Discussion by sava2021** [meta-pytorch/botorch#975] [Ideas]
Hi I would like to add artificial 2 constraints in "Constrained, Parallel, Multi-Objective BO in BoTorch with qEHVI and qParEGO" in botorch example but it didn't work, I was wondering if anyone can help me to solve multi objective multi-constraint problems.

Sava

**Comment by sdaulton** [comment]
Can you provide a bit more detail (ideally code) about what you tried and what part did not work?

**Comment by sava2021** [comment]
I have basically added 2 artificial contraints in the example but tgere was no result with the method. I was wondering if you have edited or tested an example for multi objective multi contraints with qEHVI and qParEGO approach?

Regards

Sava

Get Outlook for Android<https://aka.ms/ghei36>
________________________________

**Reply by sdaulton** [reply]
Here is simple demo with 4 outcome constraints. This adapts the constrained multi-objective BO tutorial to optimize the `WeldedBeam` problem. This should provide a good starting point. Let me know if you have any questions.

[constrained_multi_objective_bo.ipynb.txt](https://github.com/pytorch/botorch/files/7525335/constrained_multi_objective_bo.ipynb.txt)

**Comment by sava2021** [comment]
Hi Sam,

 very much appreciated for this.

I waswondering is there anyt example for multi constrained single objective solutions  like in the example of : "
Closed-loop batch, constrained BO in BoTorch with qEI and qNEI
"
at the bototrch website (BoTorch · Bayesian Optimization in PyTorch<https://botorch.org/tutorials/closed_loop_botorch_only>)?
[https://botorch.org/img/botorch.png]<https://botorch.org/tutorials/closed_loop_botorch_only>
BoTorch · Bayesian Optimization in PyTorch<https://botorch.org/tutorials/closed_loop_botorch_only>
Bayesian Optimization in PyTorch. Define a helper function that performs the essential BO step¶. The helper function below takes an acquisition function as an argument, optimizes it, and returns the batch $\{x_1, x_2, \ldots x_q\}$ along with the observed function values.
botorch.org


Regards

Sava



________________________________
Gönderen: Sam Daulton ***@***.***>
Gönderildi: 12 Kasım 2021 Cuma 09:33
Kime: pytorch/botorch ***@***.***>
Bilgi: sava2021 ***@***.***>; Author ***@***.***>
Konu: Re: [pytorch/botorch] adding more constraints in Constrained, Parallel, Multi-Objective BO in BoTorch with qEHVI and qParEGO (Discussion #975)


Here is simple demo with 4 outcome constraints. This adapts the constrained multi-objective BO tutorial to optimize the WeldedBeam problem. This should provide a good starting point. Let me know if you have any questions.

constrained_multi_objective_bo.ipynb.txt<https://github.com/pytorch/botorch/files/7525335/constrained_multi_objective_bo.ipynb.txt>

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/pytorch/botorch/discussions/975#discussioncomment-1629847>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AWOJP7PNLVFSNW6JEJTMGOLULSYKXANCNFSM5H3TI3QA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

**Reply by sdaulton** [reply]
You should be able to simply extend that single-objective tutorial notebook to use multiple constraints. The key changes will be:

1) create a problem with multiple constraints.
2) provide information about the constraints to the acquisition function. I do this in the example notebook I shared above using the `create_idxrs` function. In single objective optimization, this will look much like the constrained MC objective used by qParEGO (without the Chebyshev scalarization).
3) Evaluate performance of the methods taking the multiple constraints into account.

If you have issues or further questions, feel free to open an issue.

**Comment by sava2021** [comment]
Dear Sam,

i have worked on your welded beam problem , i have tested with other problems such as SRN, BNH, TNK and OSY. before your example i've used  on this example from tutorials too (https://botorch.org/tutorials/constrained_multi_objective_bo). however, i didnt get results on sobol, qehvi and q parego.  i havent figured out whats causing problem or my set uo  of the functions . Could you help me why i cant result on this functions? i uploaded an example function based on previous example file.
[https://botorch.org/img/botorch.png]<https://botorch.org/tutorials/constrained_multi_objective_bo>
BoTorch · Bayesian Optimization in PyTorch<https://botorch.org/tutorials/constrained_multi_objective_bo>
Constrained, Parallel, Multi-Objective BO in BoTorch with qEHVI and qParEGO¶. In this tutorial, we illustrate how to implement a constrained multi-objective (MO) Bayesian Optimization (BO) closed loop in BoTorch.
botorch.org

Kind Regards
Savas
________________________________
Gönderen: Sam Daulton ***@***.***>
Gönderildi: 12 Kasım 2021 Cuma 19:25
Kime: pytorch/botorch ***@***.***>
Bilgi: sava2021 ***@***.***>; Author ***@***.***>
Konu: Re: [pytorch/botorch] adding more constraints in Constrained, Parallel, Multi-Objective BO in BoTorch with qEHVI and qParEGO (Discussion #975)


You should be able to simply extend that single-objective tutorial notebook to use multiple constraints. The key changes will be:

  1.  create a problem with multiple constraints.
  2.  provide information about the constraints to the acquisition function. I do this in the example notebook I shared above using the create_idxrs function. In single objective optimization, this will look much like the constrained MC objective used by qParEGO (without the Chebyshev scalarization).
  3.  Evaluate performance of the methods taking the multiple constraints into account.

If you have issues or further questions, feel free to open an issue.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/pytorch/botorch/discussions/975#discussioncomment-1632888>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AWOJP7LOXCVIFEEHJYP7RWDULU5XJANCNFSM5H3TI3QA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

**Reply by sdaulton** [reply]
Hmm, I can't seem to find or load your example. Happy to take a look if you can share a notebook. I suspect the most likely culprit for "no results" is setting the reference point too close to the Pareto frontier, which makes it hard to find any feasible points (with respect to the constraints) that are better than the reference point. Setting the reference point to be further away would mitigate this. It's also worth doubling checking that the constraints are handled properly: e.g. all constraints are properly (and consistently) accounted for in the optimization and in the performance evaluation

**Comment by sava2021** [comment]
Hi Sam ,

i added  an example function into attached file. i tried to change reference point or constraint handling methods which neither of them worked. could you please take a  look at example file?




________________________________
Gönderen: Sam Daulton ***@***.***>
Gönderildi: 18 Kasım 2021 Perşembe 07:19
Kime: pytorch/botorch ***@***.***>
Bilgi: sava2021 ***@***.***>; Author ***@***.***>
Konu: Re: [pytorch/botorch] adding more constraints in Constrained, Parallel, Multi-Objective BO in BoTorch with qEHVI and qParEGO (Discussion #975)


Hmm, I can't seem to find or load your example. Happy to take a look if you can share a notebook. I suspect the most likely culprit for "no results" is setting the reference point too close to the Pareto frontier, which makes it hard to find any feasible points (with respect to the constraints) that are better than the reference point. Setting the reference point to be further away would mitigate this. It's also worth doubling checking that the constraints are handled properly: e.g. all constraints are properly (and consistently) accounted for in the optimization and in the performance evaluation

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/pytorch/botorch/discussions/975#discussioncomment-1662200>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AWOJP7IDHVVALLVZRIAAVV3UMR5FNANCNFSM5H3TI3QA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
